src/routing/bootstrap.rs

```diff
@@ -0,0 +1,653 @@
+                .await?;
+
+            let (response, sender) = self
+                .receive_join_response(relocate_payload.as_ref())
+                .await?;
+
+            match response {
+                JoinResponse::Approval {
+                    elders_info,
+                    section_chain,
+                } => {
+                    return Ok((self.node, Section::new(section_chain, elders_info)));
+                }
+                JoinResponse::Rejoin {
+                    elders_info: new_elders_info,
+                    section_key: new_section_key,
+                } => {
+                    if new_section_key == section_key {
+                        continue;
+                    }
+
+                    if new_elders_info.prefix.matches(&self.node.name()) {
+                        info!(
+                            "{} Newer Join response for our prefix {:?} from {:?}",
+                            self.node, new_elders_info, sender
+                        );
+                        elders_info = new_elders_info;
+                        section_key = new_section_key;
+                    } else {
+                        warn!(
+                            "Newer Join response not for our prefix {:?} from {:?}",
+                            new_elders_info, sender,
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    async fn send_join_requests(
+        &mut self,
+        elders_info: &EldersInfo,
+        section_key: bls::PublicKey,
+        relocate_payload: Option<&RelocatePayload>,
+    ) -> Result<()> {
+        let recipients: Vec<_> = elders_info
+            .elders
+            .values()
+            .map(Peer::addr)
+            .copied()
+            .collect();
+
+        let join_request = JoinRequest {
+            section_key,
+            relocate_payload: relocate_payload.cloned(),
+        };
+
+        info!(
+            "{} Sending {:?} to {:?}",
+            self.node, join_request, recipients
+        );
+
+        let variant = Variant::JoinRequest(Box::new(join_request));
+        let message = Message::single_src(
+            &self.node.keypair,
+            self.node.age,
+            DstLocation::Direct,
+            variant,
+            None,
+            None,
+        )?;
+
+        let _ = self.send_tx.send((message.to_bytes(), recipients)).await;
+
+        Ok(())
+    }
+
+    async fn receive_join_response(
+        &mut self,
+        relocate_payload: Option<&RelocatePayload>,
+    ) -> Result<(JoinResponse, SocketAddr)> {
+        while let Some((message, sender)) = self.recv_rx.recv().await {
+            match message.variant() {
+                Variant::BootstrapResponse(BootstrapResponse::Join {
+                    elders_info,
+                    section_key,
+                }) => {
+                    if !self.verify_message(&message, None) {
+                        continue;
```

Comment:
concerned whether this will cause a permanent wait, is the sender doesn't send a response anymore.

Reply:
Same